File: Typecast/TCHheaTable.m
//
//  TCHheaTable.m
//  Typecast
//
//  Created by David Schweinsberg on 3/12/12.
//  Copyright (c) 2012 David Schweinsberg. All rights reserved.
//

#import "TCHheaTable.h"
#import "TCDirectoryEntry.h"
#import "TCDataInput.h"

@implementation TCHheaTable

- (id)initWithDataInput:(TCDataInput *)dataInput directoryEntry:(TCDirectoryEntry *)entry
{
    self = [super init];
    if (self)
    {
        self.directoryEntry = [entry copy];
        _version = [dataInput readInt];
        _ascender = [dataInput readShort];
        _descender = [dataInput readShort];
        _lineGap = [dataInput readShort];
        _advanceWidthMax = [dataInput readShort];
        _minLeftSideBearing = [dataInput readShort];
        _minRightSideBearing = [dataInput readShort];
        _xMaxExtent = [dataInput readShort];
        _caretSlopeRise = [dataInput readShort];
        _caretSlopeRun = [dataInput readShort];
        for (int i = 0; i < 5; i++) {
            [dataInput readShort];
        }
        _metricDataFormat = [dataInput readShort];
        _numberOfHMetrics = [dataInput readUnsignedShort];
    }
    return self;
}

- (uint32_t)type
{
    return TCTable_hhea;
}

- (NSString *)description
{
    return [NSString stringWithFormat:
            @"'hhea' Table - Horizontal Header\n--------------------------------"
            @"\n        'hhea' version:       %x"  // .append(Fixed.floatValue(version))
            @"\n        yAscender:            %d"  // .append(ascender)
            @"\n        yDescender:           %d"  // .append(descender)
            @"\n        yLineGap:             %d"  // .append(lineGap)
            @"\n        advanceWidthMax:      %d"  // .append(advanceWidthMax)
            @"\n        minLeftSideBearing:   %d"  // .append(minLeftSideBearing)
            @"\n        minRightSideBearing:  %d"  // .append(minRightSideBearing)
            @"\n        xMaxExtent:           %d"  // .append(xMaxExtent)
            @"\n        horizCaretSlopeNum:   %d"  // .append(caretSlopeRise)
            @"\n        horizCaretSlopeDenom: %d"  // .append(caretSlopeRun)
            @"\n        reserved0:            0"
            @"\n        reserved1:            0"
            @"\n        reserved2:            0"
            @"\n        reserved3:            0"
            @"\n        reserved4:            0"
            @"\n        metricDataFormat:     %d"  // .append(metricDataFormat)
            @"\n        numOf_LongHorMetrics: %d",  // .append(numberOfHMetrics)
            _version,
            _ascender,
            _descender,
            _lineGap,
            _advanceWidthMax,
            _minLeftSideBearing,
            _minRightSideBearing,
            _xMaxExtent,
            _caretSlopeRise,
            _caretSlopeRun,
            _metricDataFormat,
            _numberOfHMetrics];
}

@end

File: Typecast/TCHheaTable.swift
//
//  TCHheaTable.swift
//  Type Designer
//
//  Created by David Schweinsberg on 7/14/17.
//  Copyright © 2017 David Schweinsberg. All rights reserved.
//

import Foundation
import IOUtils

class TCHheaTable: TCBaseTable, Codable {
  var version: UInt32
  var ascender: Int
  var descender: Int
  var lineGap: Int16
  var advanceWidthMax: Int16
  var minLeftSideBearing: Int16
  var minRightSideBearing: Int16
  var xMaxExtent: Int16
  var caretSlopeRise: Int16
  var caretSlopeRun: Int16
  var metricDataFormat: Int16
  var numberOfHMetrics: Int

  override init() {
    version = 0
    ascender = 0
    descender = 0
    lineGap = 0
    advanceWidthMax = 0
    minLeftSideBearing = 0
    minRightSideBearing = 0
    xMaxExtent = 0
    caretSlopeRise = 0
    caretSlopeRun = 0
    metricDataFormat = 0
    numberOfHMetrics = 0
    super.init()
  }

  init(data: Data) {
    let dataInput = TCDataInput(data: data)
    version = dataInput.readUInt32()
    ascender = Int(dataInput.readInt16())
    descender = Int(dataInput.readInt16())
    lineGap = dataInput.readInt16()
    advanceWidthMax = dataInput.readInt16()
    minLeftSideBearing = dataInput.readInt16()
    minRightSideBearing = dataInput.readInt16()
    xMaxExtent = dataInput.readInt16()
    caretSlopeRise = dataInput.readInt16()
    caretSlopeRun = dataInput.readInt16()
    for _ in 0 ..< 5 {
      _ = dataInput.readInt16()
    }
    metricDataFormat = dataInput.readInt16()
    numberOfHMetrics = Int(dataInput.readUInt16())
    super.init()
  }

  override class var tag: TCTableTag {
    get {
      return .hhea
    }
  }

  override var description: String {
    get {
      return "'hhea' Table - Horizontal Header\n--------------------------------" +
        "\n        'hhea' version:       \(version)" +  // .append(Fixed.floatValue(version))
        "\n        yAscender:            \(ascender)" +
        "\n        yDescender:           \(descender)" +
        "\n        yLineGap:             \(lineGap)" +
        "\n        advanceWidthMax:      \(advanceWidthMax)" +
        "\n        minLeftSideBearing:   \(minLeftSideBearing)" +
        "\n        minRightSideBearing:  \(minRightSideBearing)" +
        "\n        xMaxExtent:           \(xMaxExtent)" +
        "\n        horizCaretSlopeNum:   \(caretSlopeRise)" +
        "\n        horizCaretSlopeDenom: \(caretSlopeRun)" +
        "\n        reserved0:            0" +
        "\n        reserved1:            0" +
        "\n        reserved2:            0" +
        "\n        reserved3:            0" +
        "\n        reserved4:            0" +
        "\n        metricDataFormat:     \(metricDataFormat)" +
        "\n        numOf_LongHorMetrics: \(numberOfHMetrics)"
    }
  }
}
